packages/block-editor/src/components/colors-gradients/control.js

```diff
@@ -0,0 +1,160 @@
+
+const colorsAndGradientKeys = [ 'colors', 'disableCustomColors', 'gradients', 'disableCustomGradients' ];
+
+function VisualLabel( { colors, gradients, label, currentTab, colorValue, gradientValue } ) {
+	let value, ariaLabel;
+	if ( currentTab === 'color' ) {
+		if ( colorValue ) {
+			value = colorValue;
+			const colorObject = getColorObjectByColorValue( colors, value );
+			const colorName = colorObject && colorObject.name;
+			ariaLabel = sprintf( colorIndicatorAriaLabel, colorName || value );
+		}
+	} else if ( currentTab === 'gradient' && gradientValue ) {
+		value = gradientValue;
+		const gradientObject = __experimentalGetGradientObjectByGradientValue( gradients, value );
+		const gradientName = gradientObject && gradientObject.name;
+		ariaLabel = sprintf( gradientIndicatorAriaLabel, gradientName || value );
+	}
+
+	return (
+		<>
+			{ label }
+			{ !! value && (
+				<ColorIndicator
+					colorValue={ value }
+					aria-label={ ariaLabel }
+				/>
+			) }
+		</>
+	);
+}
+
+function ColorGradientControlInner( {
+	colors,
+	gradients,
+	disableCustomColors,
+	disableCustomGradients,
+	className,
+	label,
+	onColorChange,
+	onGradientChange,
+	colorValue,
+	gradientValue,
+} ) {
+	const canChooseAColor = onColorChange && ( ! isEmpty( colors ) || ! disableCustomColors );
+	const canChooseAGradient = onGradientChange && ( ! isEmpty( gradients ) || ! disableCustomGradients );
+	const [ currentTab, setCurrentTab ] = useState( gradientValue ? 'gradient' : !! canChooseAColor && 'color' );
+
+	if ( ! canChooseAColor && ! canChooseAGradient ) {
+		return null;
+	}
+	return (
+		<BaseControl
+			className={ classnames( 'block-editor-color-gradient-control', className ) }
+		>
+			<BaseControl.VisualLabel>
+				<VisualLabel
+					currentTab={ currentTab }
+					label={ label }
+					colorValue={ colorValue }
+					gradientValue={ gradientValue }
+				/>
+			</BaseControl.VisualLabel>
+			{ canChooseAColor && canChooseAGradient && (
+				<ButtonGroup className="block-editor-color-gradient-control__button-tabs">
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'color' }
+						isSecondary={ currentTab !== 'color' }
+						onClick={ () => ( setCurrentTab( 'color' ) ) }
+					>
+						{ __( 'Solid Color' ) }
+					</Button>
+					<Button
+						isLarge
+						isPrimary={ currentTab === 'gradient' }
+						isSecondary={ currentTab !== 'gradient' }
+						onClick={ () => ( setCurrentTab( 'gradient' ) ) }
+					>
+						{ __( 'Gradient' ) }
+					</Button>
+				</ButtonGroup>
+			) }
+			{ currentTab === 'color' && (
+				<ColorPalette
+					value={ colorValue }
+					onChange={ canChooseAGradient ?
+						( newColor ) => {
+							onColorChange( newColor );
+							onGradientChange();
+						} :
+						onColorChange
+					}
+					{ ... { colors, disableCustomColors } }
+				/>
+			) }
+			{ currentTab === 'gradient' && (
+				<GradientPicker
+					value={ gradientValue }
+					onChange={ canChooseAGradient ?
```

Comment:
When would a user be picking a gradient if they don't have `canChooseAGradient`? Or (judging by the symmetry with the previous tab) was this meant to be `canChooseAColor`?